development.md

```diff
@@ -0,0 +1,105 @@
+# The Compose Specification - Development support
+{:.no_toc}
+
+*Note:* Development support is an OPTIONAL part of the Compose Specification
+
+* ToC
+{:toc}
+
+## Introduction
+
+Compose specification is a platform-neutral way to define multi-container applications. A Compose implementation
+focussing on development use-case to run application on local machine will obviously also support some development hooks to improve developer velocity on his local workflow (also known as "inner loop"). The Compose Development specification allows to define how a Compose implementation SHOULD behave to efficiently assist the developer.
+
+## Definitions
+
+Compose Specification is extended to support an OPTIONAL `development` subsection on services. This section
+defines the development constraints and workflows to be set bu a Compose Implementation to assist the developer
+working on service codebase. Only a subset of Compose file services MAY define such a Development subsection.
+
+## Illustrative sample
+
+The following sample illustrates Compose specification concepts with a concrete sample application. The sample is non-normative.
+
+```yaml
+services:
+  frontend:
+    image: awesome/webapp
+    build: ./webapp
+    development:
+      watch: 
+        update: sync
+        path: src:/app
+        ignores:
+            - node_modules/
+
+  backend:
+    image: awesome/backend
+    build: ./backend
+    development:
+      watch: 
+        update: build
+      quiet-period: 5s
+
+  db:
+    image: awesome/database
+    development:
+      watch: sighup
+```
+
+## Developemnt mode definition
+
+The `dev` element define configuration options that are applied by Compose implementations to assist developer during development of a service.
+
+### watch
+
+`watch` can be specified either as a string containing policy or a detailed structure/
+
+```yml
+services:
+  webapp:
+    development: 
+      watch: build
+```
+is equivalent to 
+
+```yml
+services:
+  webapp:
+    development: 
+      watch: 
+        policy: build
+```
+
+
+The `watch` attribute defines `update` policy to update a running service as Compose Implementation detects 
+changes in service source code. "Source code" to watch is identified by the `build.context` section.
+
+While Compose implementations can support custom values, the specification defines:
+- `build`: rebuild service image based on the `build` section and restart the service with updated image
+- `sync`: synchronize source files with container content. `path` attribute is REQUIRED to define the mapping 
```

Comment:
Could you elaborate a bit more on how this interacts with host mounts in `volumes`?

Reply:
This is intentionally unspecified so implementation can be flexible. Most obvious approach is to add an implicit bind mount for sources. For illustration, a classic nodeJS application uses a COPY directive in Dockerfile to add sources into /app folder. With such a sync directive, `docker compose dev` would add a new bind mount so that local sources override this /app folder and get synced inside container, assuming node runtime manages hot-reload.

Reply:
Thank you for explaining 👍 My concern is that we are potentially adding another way to mount files into the containers to the spec without necessarily drawing a clear distinction for when users should do volumes and when they should be using `develop` section instead. Is there a use case that this would cover that volumes are not able to?

Reply:
This is still unclear for me.